arm_robots/src/arm_robots/cartesian.py

```diff
@@ -0,0 +1,221 @@
+
+
+def point_to_np_array(a):
+    return np.array([a.x, a.y, a.z])
+
+
+def pose_distance(a, b):
+    return np.linalg.norm(point_to_np_array(a.position) - point_to_np_array(b.position))
+
+
+class ArmSide(enum.IntEnum):
+    LEFT = 0
+    RIGHT = 1
+
+
+class CartesianImpedanceController:
+    def __init__(self, tf_buffer, motion_status_listeners, motion_command_publisher, joint_lim_low, joint_lim_high,
+                 position_close_enough=0.0025, timeout_per_1m=500, intermediate_acceptance_factor=7.,
+                 joint_limit_boundary=0.03):
+        """
+
+        :param tf_buffer: tf2 Buffer object
+        :param motion_status_listeners: ROS listener (wrapper around subscriber) for status messages
+        :param motion_command_publisher: ROS publisher for arm commands
+        :param joint_lim_low: lower joint limits in degrees
+        :param joint_lim_high: upper joint limits in degrees
+        :param position_close_enough: Distance (m) to target position to be considered close enough
+        :param timeout_per_1m: Allowed time (s) to execute before timing out per 1m of travel
+        :param joint_limit_boundary: Angle (radian or list of radian) boundary of each joint limit to avoid by
+        returning to the previous pose for any entering. If this boundary is larger than what any single motion command
+        will step, then we will not receive exceptions on the robot side.
+        """
+        self.target_pose = None
+        # for users to read after reaching goal
+        self.reached_joint_limit = False
+        self.timed_out = False
+        # joint limits
+        self.joint_lim_low = joint_lim_low
+        self.joint_lim_high = joint_lim_high
+
+        # tf
+        self.tf_buffer = tf_buffer
+        # what frames the measured cartesian pose is given in
+        self.sensor_frames = ["victor_left_arm_world_frame_kuka", "victor_right_arm_world_frame_kuka"]
+        self.world_frame = "victor_root"
+
+        # goal parameters
+        self._intermediate_target = None
+        self._dists_to_goal = []
+        self.position_close_enough = position_close_enough
+        self._intermediate_close_enough = position_close_enough * intermediate_acceptance_factor
+        self._this_target_start_time = None
+        self._goal_start_time = None
+        self._init_goal_dist = None
+        self._timeout_per_1m = timeout_per_1m
+
+        # safety parameters
+        self._joint_boundary = joint_limit_boundary
+        if isinstance(self._joint_boundary, float):
+            self._joint_boundary = [self._joint_boundary for _ in self.joint_lim_low]
+        assert len(self._joint_boundary) == len(self.joint_lim_low)
+        self._joint_boundary = np.array(self._joint_boundary)
+        self._intermediate_target_start_pose = None
+        self._check_joint_limits = True
+        self._start_violation = 0
+
+        self.active_arm = ArmSide.LEFT
+        # ROS subscribers
```

Comment:
these three comments seem pretty pointless, and the last one even seems misleading? is there a control loop somewhere I'm missing?